Common: make BitField endianness-aware

wwylele opened the pull request:
With this `BitField` now has storage specification on byte order just like `u32_le`/`u32_be`. The original `BitField` is default to little endian, while the big-endian counterpart is aliased to `BitFieldBE`. This resolves the problem of `BitField<*, *, u32_le>` not compilable on BE machines.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4599)
<!-- Reviewable:end -->

zhaowenlan1779 commented on src/common/swap.h:
```diff
@@ -172,7 +172,7 @@ struct swap_struct_t {
 
 protected:
-    T value = T();
+    T value;
```
Ah, sorry, didn't see it

wwylele commented:
Update: simplified template specialization for `AddEndian<>`:
 - unified all `AddEndian<T, KeepTag>` to a single template that forwards the original type
 - implemented `AddEndian<enum, SwapTag>` as a less-specialized specialization with `static_assert` inside
